src/cryptonote_core/tx_pool.cpp

```diff
@@ -900,6 +899,86 @@ namespace cryptonote
     }, false, category);
   }
   //------------------------------------------------------------------
+  bool tx_memory_pool::get_pool_info(time_t start_time, bool include_sensitive, std::vector<tx_details>& added_txs, std::vector<crypto::hash>& removed_txs, bool& incremental) const
+  {
+    CRITICAL_REGION_LOCAL(m_transactions_lock);
+    incremental = true;
+    if (start_time == (time_t)0)
+    {
+      // Giving no start stime means give back whole pool
+      incremental = false;
+    }
+    else if ((m_added_txs_start_time != (time_t)0) && (m_removed_txs_start_time != (time_t)0))
+    {
+      if ((start_time <= m_added_txs_start_time) || (start_time <= m_removed_txs_start_time))
+      {
+        // If either of the two lists do not go back far enough it's not possible to
+        // deliver incremental pool info
+        incremental = false;
+      }
+      // The check uses "<=": We cannot be sure to have ALL txs exactly at start_time, only AFTER that time
+    }
+    else
+    {
+      // Some incremental info still missing completely
+      incremental = false;
+    }
+
+    added_txs.clear();
+    removed_txs.clear();
+
+    if (!incremental)
+    {
+      // Give back the whole pool in 'added_txs'; because 'get_transaction_info' makes itself a transaction
+      // we have to get a list first and then get the infos
+      std::vector<crypto::hash> txids;
+      {
+        CRITICAL_REGION_LOCAL1(m_blockchain);
```

Comment:
I see your point of potential divergences by not locking over the whole process, but I am not sure what you propose.

In a first version of the code I had the `get_transaction_info` call directly in the anonymous method for `for_all_txpool_txes`, and that gave errors with LMDB complaining with some mysterious (to me) error about transactions nested in a wrong way somehow. Locking only for `for_all_txpool_txes` and then calling `for_all_txpool_txes` outside solved the problem.

If you propose to change the locking behaviour of `get_transaction_info` to be able to lock over the whole process here: I am not sure how simple that will get because the call is used in other places as well.

Reply:
Maybe we talk past each other somewhat here:

`get_transaction_info` as it is now includes this at the start:

    CRITICAL_REGION_LOCAL(m_transactions_lock);
    CRITICAL_REGION_LOCAL1(m_blockchain);

I strongly suspect that you can't nest those "critical reasons". When I did in a slightly different way that you propose I got quite mysterious-sounding errors about LMDB transactions nested in a wrong way somehow, as already mentioned. With your proposal and an unmodified `get_transaction_info` nesting would occur, and the errors would probably return.

Of course we could make a second private version of `get_transaction_info` without the locking to use it here and enable us to lock over the whole method execution. Do you think the effort would be worth it?

Reply:
I meant add `CRITICAL_REGION_LOCAL1(m_blockchain)` at the top of `get_pool_info`

Reply:
Yes, I already understood your proposal as such, but I am afraid I did not yet manage to get accross what's the problem with that :)

With that we will have a lock that is active during the whole execution of the `get_pool_info` method, right? The look will go away when its destructor gets called invisibly at the very end of the method, so to say.

That means that when the method makes calls to `get_transaction_info` sooner or later, the lock will be active - of course. Now, what do you see if you switch over to the code of `get_transaction_info`? Right, that method **also** sets a lock, the same type of lock if I am not gravely mistaken.

I stronlgy suspect such "nesting" of locks does **not** work. When I tried to nest (in a slightly different way), I got LMDB errors about transactions not properly nested.

See what I mean?

The solution is not too hard, however, seems to it simply needs two versions of `get_transaction_info`, one without any locks (can be private, for use by `get_pool_info`) and a second, public one that sets locks and then also calls the private one for the real work.

My question to you was: Do you thing the problem at hand is important enough to make that modification? I guess the answer is "yes" ...

Reply:
It should work fine because the underlying mutex is reentrant (I recently realized this too), and I would say it's worth it because it protects against races/inconsistent states that will be difficult to track

Reply:
Done. The problem I insisted on so vehemently is real (I just tried again) but must have another reason than I assumed, the nesting of the locks. For some reason still unknown to me it's indeed not possible to directly call `get_transaction_info` inside the anonymous method used for `for_all_txpool_txes`, one has to collect the ids first and then get the infos afterwards. But which is now no problem anymore thanks to moving the lock to the start of the method, as you proposed.